### PR TITLE
radosgw systemd support

### PR DESCRIPTION
**Pull request by osynge**
Added a radosgw systemd support and associated prestart script.
  - With improved checking over first revison.
  - ceph-radosgw-prestart.sh now installed in /usr/lib/ceph-radosgw

Signed-off-by: Owen Synge <osynge@suse.com>

### CODE REVIEW

**Comment by liewegas** [comment]
This seems more complicated than what we want.  It seems like we should be modeling this after the ceph-mds service, which is pretty simple.  The goal is for a static (and perhaps eventually empty) ceph.conf that doesn't enumerate daemons.  In hammer, for instance, 'ceph-deploy rgw create HOST' will create a client.rgw.$hostname instance in /var/lib/ceph/radosgw/ceph-rgw.$hostname.  Since it's systemd there is the additional 'enable' step (which we don't do yet, but should).

What do you think?

**Comment by osynge** [comment]
I agree its more complex than I wanted. I suspect when you move to the new radosgw codebase not based on fastcgi things will be better. Currently to my understanding the ceph.conf is needed to be setup. I look forward to this not being needed. Systemd is rather inflexable about gathering variables at runtime though.

I have 2 reasons for this -> apart form my concerns that systemd does lead to over complication.

1) Supporting ceph-radosgw@.service allows users to enable multiple radosgw services on a single host,

2) Since its paramterised, we have had issues with people setting up radosgw with invalid paramters.

In summary with people setting things up by hand we had many bug reports (from our team not even customers) due to misconfiguration. I wanted to validate these requests as much as possible as I belive in fail fast.

the forthcoming patch to ceph-deploy to support radosgw setting up apache and fastcgi sets all this up and enables radosgw.

If thier is a way to set up radosgw without ceph.conf configuration I would be interested and keen to remove the code.

File: systemd/ceph-radosgw@.service [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[Unit]
+Description=Ceph rados gateway
+After=network-online.target local-fs.target
+Wants=network-online.target local-fs.target
+PartOf=ceph.target
+
+[Service]
+EnvironmentFile=-/etc/sysconfig/ceph
+Environment=CLUSTER=ceph
+ExecStart=/usr/bin/radosgw -f --conf /etc/ceph/${CLUSTER}.conf --name client.radosgw.%i
```
Comment (liewegas) [inline systemd/ceph-radosgw@.service]: --cluster ${CLUSTER}, and all the other substitutions will happen on their own

File: src/ceph-radosgw-prestart.sh [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  exit 1
+fi
+
+# mapped to this host?
+host=`${CEPHCONF} -c ${ceph_conf_file} -n $longname host`
+hostname=`hostname -s`
+if [ "$host" != "$hostname" ]; then
+  echo "hostname $hostname could not be found in ceph.conf:[$longname], not starting."
+  exit 1
+fi
```
Comment (liewegas) [inline src/ceph-radosgw-prestart.sh]: this host mapping piece is not necessary

File: systemd/ceph-radosgw@.service (hunk 1 above)
Comment (liewegas) [inline systemd/ceph-radosgw@.service]: can we standardize on client.rgw.$id ?  at some point i'd even like to rename radosgw -> ceph-rgw for simplicity.

Comment (osynge) [inline systemd/ceph-radosgw@.service]: sure -> I am happy with this substitution.

I felt it was cleaner to have ${CLUSTER} as part of the environment, are you saying that if the environment variable is specified  "--conf /etc/ceph/${CLUSTER}.conf" is not needed?

**Comment by osynge** [comment]
Im not sure this code is suitable for master, and is only for firefly now, as I have not looked into the new rgw, if its suitable for both thats great, but I have not yet tested the new rgw yet.

File: systemd/ceph-radosgw@.service (hunk 1 above)
Comment (liewegas) [inline systemd/ceph-radosgw@.service]: I'm saying if you do "-C ${CLUSTER}" ceph will substitute it in in all the 
other places (like the default config file /etc/ceph/$cluster.conf, but 
also lots of other places).

**Comment by dachary** [comment]
@osynge would you like to close this pull request and open another against firefly maybe ?

**Comment by osynge** [comment]
Since I dont yet have good master testing infrastructure I am not sure. I have a few more bugs in my intray before Im back on the investing in testing / patch submission maintenance cycle.

**Comment by dachary** [comment]
@osynge understood :-) Shall we close this one for now then ?

**Comment by osynge** [comment]
ok

File: src/ceph-radosgw-prestart.sh (hunk 2 above)
Comment (osynge) [inline src/ceph-radosgw-prestart.sh]: Your right, this cant be done when you want to work on an alias.

**Comment by osynge** [comment]
trying to reopen to debug.

File: systemd/ceph-radosgw@.service (hunk 1 above)
Comment (osynge) [inline systemd/ceph-radosgw@.service]: OK I did this. So sorry for the delay.

**Comment by liewegas** [comment]
closing this in fvor of wip-systemd
